nomos-services/blend/src/core/mod.rs

```diff
@@ -735,42 +703,58 @@ where
-                recovery_checkpoint = handle_local_data_message(local_data_message, &mut crypto_processor, message_scheduler, recovery_checkpoint).await;
+                recovery_checkpoint = handle_local_data_message(local_data_message, &mut crypto_processor, &mut message_scheduler, recovery_checkpoint).await;
             }
             Some(incoming_message) = blend_messages.next() => {
-                recovery_checkpoint = handle_incoming_blend_message(incoming_message, message_scheduler, &crypto_processor, old_session_crypto_processor.as_ref(), blending_token_collector, recovery_checkpoint);
+                recovery_checkpoint = handle_incoming_blend_message(incoming_message, &mut message_scheduler, old_session_message_scheduler.as_mut(), &crypto_processor, old_session_crypto_processor.as_ref(), blending_token_collector, recovery_checkpoint);
             }
             Some(round_info) = message_scheduler.next() => {
                 recovery_checkpoint = handle_release_round(round_info, &mut crypto_processor, rng, backend, network_adapter, recovery_checkpoint).await;
             }
+            Some(processed_messages_to_release) = async {
+                if let Some(old_scheduler) = &mut old_session_message_scheduler {
+                    old_scheduler.next().await
+                } else {
+                    None
+                }
+            } => {
+                recovery_checkpoint = handle_release_round_for_old_session(processed_messages_to_release, rng, backend, network_adapter, recovery_checkpoint).await;
+            }
             Some(clock_tick) = remaining_clock_stream.next() => {
                 public_info = handle_clock_event(clock_tick, blend_config, epoch_handler, &mut crypto_processor, backend, public_info).await;
             }
             Some(pol_info) = secret_pol_info_stream.next() => {
                 handle_new_secret_epoch_info(pol_info.poq_private_inputs, &mut crypto_processor);
             }
             Some(session_event) = remaining_session_stream.next() => {
-                match handle_session_event(session_event, blend_config, crypto_processor, public_info, recovery_checkpoint, blending_token_collector, backend).await {
-                    HandleSessionEventOutput::Transitioning { new_crypto_processor, old_crypto_processor, new_public_info, new_recovery_checkpoint } => {
+                match handle_session_event(session_event, blend_config, crypto_processor, message_scheduler, public_info, recovery_checkpoint, blending_token_collector, backend, rng.clone()).await {
+                    HandleSessionEventOutput::Transitioning { new_crypto_processor, old_crypto_processor, new_scheduler, old_scheduler, new_public_info, new_recovery_checkpoint } => {
                         crypto_processor = new_crypto_processor;
                         old_session_crypto_processor = Some(old_crypto_processor);
+                        message_scheduler = new_scheduler;
```

Comment:
I guess the same comment as for the crypto processor applies here: why do we need to keep track of two schedulers instead of having all the logic internally, as we do for, for instance, our libp2p swarm and behaviours? Or our proof verifiers?